File: server/src/co2Reader.ts
/**
 * CO2 Reader Module
 *
 * Handles USB HID communication with TFA Dostmann AirControl Mini CO2 Meter.
 * Device: TFA 31.5006
 * Vendor ID: 0x04D9
 * Product ID: 0xA052
 */

import HID from 'node-hid';

// Device identifiers
const VENDOR_ID = 0x04d9;
const PRODUCT_ID = 0xa052;

// XOR encryption key (used by the device)
const XOR_KEY = new Uint8Array([
  0x86, 0x41, 0xc9, 0xa8, 0x7f, 0x41, 0x3c, 0xac
]);

export interface CO2Reading {
  co2Ppm: number | null;
  temperatureCelsius: number | null;
  timestamp: Date;
}

export interface DeviceInfo {
  product: string;
  vendorId: string;
  productId: string;
  path: string;
}

/**
 * Decrypt data from the CO2 sensor using XOR cipher.
 */
function decrypt(data: Buffer): Buffer {
  const result = Buffer.alloc(8);
  const shuffle = [2, 4, 0, 7, 1, 6, 5, 3];

  // Phase 1: Shuffle
  for (let i = 0; i < 8; i++) {
    result[shuffle[i]] = data[i];
  }

  // Phase 2: XOR with key
  for (let i = 0; i < 8; i++) {
    result[i] ^= XOR_KEY[i];
  }

  // Phase 3: Shift operations
  const temp = Buffer.alloc(8);
  for (let i = 0; i < 8; i++) {
    temp[i] = ((result[i] >> 3) | (result[(i + 7) % 8] << 5)) & 0xff;
  }

  // Phase 4: Subtract with magic number
  const magic = 0x23;
  for (let i = 0; i < 8; i++) {
    temp[i] = (temp[i] - magic) & 0xff;
  }

  return temp;
}

/**
 * List all connected CO2 devices.
 */
export function listDevices(): DeviceInfo[] {
  const devices = HID.devices();
  return devices
    .filter(d => d.vendorId === VENDOR_ID && d.productId === PRODUCT_ID)
    .map(d => ({
      product: d.product || 'CO2 Monitor',
      vendorId: `0x${d.vendorId.toString(16).padStart(4, '0').toUpperCase()}`,
      productId: `0x${d.productId.toString(16).padStart(4, '0').toUpperCase()}`,
      path: d.path || ''
    }));
}

/**
 * CO2 Reader class for reading from the sensor.
 */
export class CO2Reader {
  private device: HID.HID | null = null;
  private lastReading: CO2Reading = {
    co2Ppm: null,
    temperatureCelsius: null,
    timestamp: new Date()
  };

  /**
   * Connect to the CO2 sensor.
   */
  connect(): boolean {
    try {
      const devices = listDevices();
      if (devices.length === 0) {
        console.error('No CO2 device found');
        return false;
      }

      this.device = new HID.HID(VENDOR_ID, PRODUCT_ID);

      // Send initialization command
      const initCmd = Buffer.alloc(9);
      initCmd[0] = 0x00; // Report ID
      this.device.sendFeatureReport(Array.from(initCmd));

      console.log(`Connected to CO2 sensor: ${devices[0].product}`);
      return true;
    } catch (error) {
      console.error('Failed to connect to CO2 sensor:', error);
      return false;
    }
  }

  /**
   * Disconnect from the sensor.
   */
  disconnect(): void {
    if (this.device) {
      this.device.close();
      this.device = null;
      console.log('Disconnected from CO2 sensor');
    }
  }

  /**
   * Check if connected to the sensor.
   */
  isConnected(): boolean {
    return this.device !== null;
  }

  /**
   * Read data from the sensor.
   * Returns the latest reading after processing available data.
   */
  read(timeoutMs: number = 5000): CO2Reading {
    if (!this.device) {
      return this.lastReading;
    }

    const startTime = Date.now();
    let co2 = this.lastReading.co2Ppm;
    let temp = this.lastReading.temperatureCelsius;

    try {
      while (Date.now() - startTime < timeoutMs) {
        const data = this.device.readTimeout(100);

        if (!data || data.length < 8) {
          continue;
        }

        const buffer = Buffer.from(data);
        const decrypted = decrypt(buffer);

        // Validate checksum
        const checksum = (decrypted[0] + decrypted[1] + decrypted[2]) & 0xff;
        if (checksum !== decrypted[3]) {
          continue;
        }

        const op = decrypted[0];
        const value = (decrypted[1] << 8) | decrypted[2];

        // CO2 reading (op = 0x50)
        if (op === 0x50) {
          co2 = value;
        }
        // Temperature reading (op = 0x42)
        else if (op === 0x42) {
          temp = value / 16.0 - 273.15;
          temp = Math.round(temp * 10) / 10; // Round to 1 decimal
        }

        // If we have both readings, return
        if (co2 !== null && temp !== null) {
          this.lastReading = {
            co2Ppm: co2,
            temperatureCelsius: temp,
            timestamp: new Date()
          };
          return this.lastReading;
        }
      }
    } catch (error) {
      console.error('Error reading from sensor:', error);
    }

    // Return whatever we have
    if (co2 !== null || temp !== null) {
      this.lastReading = {
        co2Ppm: co2,
        temperatureCelsius: temp,
        timestamp: new Date()
      };
    }

    return this.lastReading;
  }

  /**
   * Get the last reading without blocking.
   */
  getLastReading(): CO2Reading {
    return this.lastReading;
  }
}

// Test if run directly
if (import.meta.url === `file://${process.argv[1]}`) {
  console.log('CO2 Reader Test');
  console.log('===============');

  const devices = listDevices();
  console.log(`Found ${devices.length} device(s):`, devices);

  if (devices.length > 0) {
    const reader = new CO2Reader();
    if (reader.connect()) {
      console.log('Reading from sensor...');
      const reading = reader.read();
      console.log(`CO2: ${reading.co2Ppm} ppm`);
      console.log(`Temperature: ${reading.temperatureCelsius}°C`);
      reader.disconnect();
    }
  }
}

File: co2_reader.py
"""
CO2 Reader Module for TFA Dostmann AirControl Mini (31.5006)

Reads CO2 concentration and temperature from the USB HID device.
Based on reverse-engineered protocol from:
- https://github.com/JsBergbau/TFACO2AirCO2ntrol_CO2Meter
- https://github.com/MathieuSchopfer/tfa-airco2ntrol-mini
"""

import hid
import time
from dataclasses import dataclass
from typing import Optional

# Device identifiers
VENDOR_ID = 0x04D9
PRODUCT_ID = 0xA052

# Data type codes
CO2_CODE = 0x50  # 80 decimal
TEMP_CODE = 0x42  # 66 decimal

# Encryption key (sent to device to initialize)
KEY = bytes([0xC4, 0xC6, 0xC0, 0x92, 0x40, 0x23, 0xDC, 0x96])


@dataclass
class Reading:
    """A single reading from the CO2 monitor."""
    co2_ppm: Optional[int] = None
    temperature_celsius: Optional[float] = None
    timestamp: float = 0.0

    def is_complete(self) -> bool:
        """Check if both CO2 and temperature have been read."""
        return self.co2_ppm is not None and self.temperature_celsius is not None


class CO2Reader:
    """Reader for TFA Dostmann AirControl Mini CO2 Monitor."""

    def __init__(self):
        self.device: Optional[hid.device] = None
        self._connected = False

    def connect(self) -> bool:
        """Connect to the CO2 monitor device."""
        try:
            self.device = hid.device()
            self.device.open(VENDOR_ID, PRODUCT_ID)
            self.device.set_nonblocking(False)

            # Send key to initialize device (feature report)
            # Report ID 0x00 followed by the key
            self.device.send_feature_report(bytes([0x00]) + KEY)

            self._connected = True
            return True
        except IOError as e:
            print(f"Failed to connect to CO2 monitor: {e}")
            self._connected = False
            return False

    def disconnect(self):
        """Disconnect from the device."""
        if self.device:
            try:
                self.device.close()
            except Exception:
                pass
        self.device = None
        self._connected = False

    def is_connected(self) -> bool:
        """Check if device is connected."""
        return self._connected and self.device is not None

    def _decrypt(self, data: bytes) -> bytes:
        """Decrypt data from device using XOR with key."""
        # The device uses a simple XOR encryption with shuffled bytes
        # Shuffle order: [2, 4, 0, 7, 1, 6, 5, 3]
        shuffle = [2, 4, 0, 7, 1, 6, 5, 3]
        phase1 = bytes([data[shuffle[i]] for i in range(8)])

        # XOR with key
        phase2 = bytes([phase1[i] ^ KEY[i] for i in range(8)])

        # Final transformation
        result = bytes([
            ((phase2[i] >> 3) | (phase2[(i - 1) % 8] << 5)) & 0xFF
            for i in range(8)
        ])

        return result

    def _parse_value(self, data: bytes) -> tuple[int, int]:
        """Parse decrypted data to extract type and value."""
        # First byte is the data type
        data_type = data[0]
        # Value is in bytes 1-2 (big endian)
        value = (data[1] << 8) | data[2]
        return data_type, value

    def read(self, timeout_seconds: float = 5.0) -> Reading:
        """
        Read CO2 and temperature from the device.

        Args:
            timeout_seconds: Maximum time to wait for complete reading

        Returns:
            Reading object with CO2 and temperature values
        """
        if not self.is_connected():
            raise IOError("Device not connected")

        reading = Reading(timestamp=time.time())
        start_time = time.time()

        while not reading.is_complete():
            if time.time() - start_time > timeout_seconds:
                break

            try:
                # Read 8 bytes from device
                data = self.device.read(8, timeout_ms=1000)
                if not data or len(data) < 8:
                    continue

                # Decrypt the data
                decrypted = self._decrypt(bytes(data))

                # Validate checksum (byte 3 should equal byte 0 + byte 1 + byte 2)
                checksum = (decrypted[0] + decrypted[1] + decrypted[2]) & 0xFF
                if checksum != decrypted[3]:
                    # Try without decryption (newer firmware)
                    data_type, value = self._parse_value(bytes(data))
                else:
                    data_type, value = self._parse_value(decrypted)

                # Parse based on data type
                if data_type == CO2_CODE:
                    reading.co2_ppm = value
                elif data_type == TEMP_CODE:
                    # Temperature formula: value / 16.0 - 273.15
                    reading.temperature_celsius = round(value / 16.0 - 273.15, 1)

            except IOError:
                # Device may have disconnected
                self._connected = False
                break

        return reading

    def __enter__(self):
        """Context manager entry."""
        self.connect()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.disconnect()
        return False


def list_devices():
    """List all compatible CO2 monitor devices."""
    devices = hid.enumerate(VENDOR_ID, PRODUCT_ID)
    return devices


if __name__ == "__main__":
    # Test reading
    print("Searching for CO2 monitor...")
    devices = list_devices()

    if not devices:
        print("No CO2 monitor found!")
        print(f"Looking for VID: 0x{VENDOR_ID:04X}, PID: 0x{PRODUCT_ID:04X}")
        exit(1)

    print(f"Found {len(devices)} device(s)")

    with CO2Reader() as reader:
        if reader.is_connected():
            print("Connected! Reading data...")
            reading = reader.read()
            print(f"CO2: {reading.co2_ppm} ppm")
            print(f"Temperature: {reading.temperature_celsius} °C")
        else:
            print("Failed to connect to device")
